### PR TITLE
Fix Generator:MicroCHP allocation problem

### LINKED ISSUE #7853: Generator:MicroCHP bug results in only the last one being operable
Issue overview
--------------
User file with 2 Generator:MicroCHP objects with FollowThermal only ran one generator, the other never ran.
Some debugging revealed the problem [here](https://github.com/NREL/EnergyPlus/blob/7e32c807a764b2b6f55fc573e75f97ad8b132997/src/EnergyPlus/GeneratorDynamicsManager.cc#L165-L167), the deallocate obliterates any prior generators, so only the last one remains in the `GeneratorDynamics` array. Looks like this was originally programmed to expect a loop inside this function, but now is called multiple times, so the elseif block should be deleted here? Does this function get called for any other generator type?

I don't see any possible workaround except to use a single Generator:MicroCHP.

### Details
Some additional details for this issue (if relevant):
 - Version of EnergyPlus 9.3.0-IDDFreeze
 - Helpdesk ticket number 15401

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added to EnergyPlusDevSupport\DefectFiles
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by mjwitte**
Pull request overview
---------------------
 - Fixes #7853
 - Fix allocation bug that prevents more than one Generator:MicroCHP from operating.

### Review Notes
 - First commit adds unit test, EnergyPlusFixture.MicroCHPTest_InitGeneratorDynamics, which [fails](https://nrel.github.io/EnergyPlusBuildResults/EnergyPlus-345f829b7991a5150495189943b1c0d28dc46c81-x86_64-Linux-Ubuntu-18.04-gcc-7.5-UnitTestsCoverage-Debug.html).
 - Next commit adds the fix, new unit test add [10% coverage to GeneratorDynamics.cc](http://energyplus.s3-website-us-east-1.amazonaws.com/coverage/2020-03-12-EnergyPlus-06f0e37bea-x86_64-gcc-7.5UTCD/EnergyPlus/GeneratorDynamicsManager.cc.gcov.html) which was previously 0%.
 - No diffs expected in test suite.
 - Defect file now shows both generators operating, before fix, only one operated.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [x] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [x] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [x] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [x] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [x] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [x] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] ~If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label~
 - [ ] ~If structural output changes, add to output rules file and add OutputChange label~
 - [ ] ~If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies~

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] ~If feature, test running new feature, try creative ways to break it~
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] ~Check any new function arguments for performance impacts~
 - [ ] ~Verify IDF naming conventions and styles, memos and notes and defaults~
 - [ ] ~If new idf included, locally check the err file and other outputs~

### CODE REVIEW

**Comment by mjwitte** [comment]
New unit test fails as expected before fixing the bug.

File: src/EnergyPlus/GeneratorDynamicsManager.cc [hunk 1]
```diff
@@ -137,39 +108,16 @@ namespace GeneratorDynamicsManager {
...
-        // SUBROUTINE LOCAL VARIABLE DECLARATIONS:
 
         // get the number of generators that might use this module
         NumGensWDynamics = MicroCHPElectricGenerator::NumMicroCHPs; // TODO  + NumFuelCellCGenerators
 
         if (!allocated(GeneratorDynamics)) {
             GeneratorDynamics.allocate(NumGensWDynamics);
-        } else {
-            GeneratorDynamics.deallocate();
-            GeneratorDynamics.allocate(NumGensWDynamics);
```
Comment (mjwitte) [inline src/EnergyPlus/GeneratorDynamicsManager.cc:167]: This is the fix. This function is called multiple times, so it should only allocate `GeneratorDynamics` once. These lines were clearing the data for prior generators, so only the last generator would be populated.

**Review by Myoldmopar — APPROVED** [review]
Great fix, and great unit test.  I'm running locally, but this should be an easy one to drop in right away.

**Comment by Myoldmopar** [comment]
CI is also very happy, and yes, great work on the improved unit test coverage!

**Comment by Myoldmopar** [comment]
Ran the defect file locally with develop and this branch.  In develop, I see the first generator is not running, but in this branch I see both running.  This is good to go.  Merging.

**Pull request merged**